contracts/utils/NoncesSemiAbstracted.sol

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {Nonces} from "./Nonces.sol";
+
+/**
+ * @dev Alternative to {Nonces}, that support key-ed nonces.
+ *
+ * Follows the https://eips.ethereum.org/EIPS/eip-4337#semi-abstracted-nonce-support[ERC-4337's semi-abstracted nonce system].
+ */
+abstract contract NoncesSemiAbstracted is Nonces {
```

Comment:
"Semi-abstracted" refers to the fact that ERC-4337 handles some of the nonce logic, i.e., it does not accept arbitrary nonce logic defined by accounts:
>This approach maintains the guarantee of UserOperation hash uniqueness on-chain on the protocol level while allowing wallets to implement any custom logic they may need operating on a 192-bit “key” field, while fitting the 32 byte word.

It does not refer to the key-sequence separation. A better name for this library might be `NoncesKeyed`.